Use util::Result in for calculating mempool ancestors

stickies-v opened the pull request:
Upon reviewing the documentation for `CTxMemPool::CalculateMemPoolAncestors`, I noticed `setAncestors` was meant to be an `out` parameter but actually is an `in,out` parameter, as can be observed by adding `assert(setAncestors.empty());` as the first line in the function and running `make check`. This PR fixes this unexpected behaviour and introduces refactoring improvements to make intents and effects of the code more clear.

## Unexpected behaviour
This behaviour occurs only in the package acceptance path, currently only triggered by `testmempoolaccept` and `submitpackage` RPCs. 

In `MemPoolAccept::AcceptMultipleTransactions()`, we first call `PreChecks()` and then `SubmitPackage()` with the same `Workspace ws` reference. `PreChecks` leaves `ws.m_ancestors` in a potentially non-empty state, before it is passed on to `MemPoolAccept::SubmitPackage`. `SubmitPackage` is the only place where `setAncestors` isn't guaranteed to be empty before calling `CalculateMemPoolAncestors`. The most straightforward fix is to just forcefully clear `setAncestors` at the beginning of CalculateMemPoolAncestors, which is done in the first bugfix commit.

## Improvements
### Return value instead of out-parameters
This PR updates the function signatures for `CTxMemPool::CalculateMemPoolAncestors` and `CTxMemPool::CalculateAncestorsAndCheckLimits` to use a `util::Result` return type and eliminate both the `setAncestors` `in,out`-parameter as well as the error string. It simplifies the code and makes the intent and effects more explicit.

### Observability
There are 7 instances where we currently call `CalculateMemPoolAncestors` without actually checking if the function succeeded because we assume that it can't fail, such as in [miner.cpp](https://github.com/bitcoin/bitcoin/blob/69b10212ea5370606c7a5aa500a70c36b4cbb58f/src/node/miner.cpp#L399). This PR adds a new wrapper `AssumeCalculateMemPoolAncestors` function that logs such unexpected failures, or in case of debug builds even halts the program. It's not crucial to the objective, more of an observability improvement that seems sensible to add on here.

stickies-v committed:
mempool: use util::Result for CalculateAncestorsAndCheckLimits

Avoid using setAncestors outparameter, simplify function signatures
and avoid creating unused dummy strings.

stickies-v committed:
mempool: use util::Result for CalculateMemPoolAncestors

Avoid using setAncestors outparameter, simplify function signatures
and avoid creating unused dummy strings.

w0xlt reviewed (commented):
Approach ACK

glozow reviewed (commented):
Concept ACK, definitely prefer this over out-params.

theStack commented:
Concept ACK

aureleoules reviewed (approved):
ACK a425537fbfc20d87c325b5ae2f06f2597a31cf0e - LGTM and no behavior change

furszy commented on src/node/miner.cpp:
```diff
@@ -394,9 +395,7 @@ void BlockAssembler::addPackageTxs(const CTxMemPool& mempool, int& nPackagesSele
-        CTxMemPool::setEntries ancestors;
-        std::string dummy;
-        mempool.CalculateMemPoolAncestors(*iter, ancestors, CTxMemPool::Limits::NoLimits(), dummy, false);
+        auto ancestors{*Assert(mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false))};
```
Even when this currently cannot fail (because of the no limits and search parents args), I would be more conservative and catch + log any possible result error instead of assert the result.

In this way, we keep the current behavior, and use an empty `ancestors` set instead of aborting in production if something gets changed inside `CalculateMemPoolAncestors` and we forget to update this call.

stickies-v replied:
Interesting. I see your point re robustness, but generally, I really don't like silent failures - even when logged. I'd much rather they throw so we can more easily detect and fix unexpected behaviour as early on as possible. There are 7 instances in this PR where we do this kind of assertion - would you only want to update it here or also in the other instances? I also understand that the current behaviour is that of silent failure, so I wouldn't want to break anything with this change - but as you mention, I don't think it does, given the `NoLimits()` parameter in all of these instances.

Cc'ing the other reviewers in case anyone else has a view (@w0xlt @glozow @theStack @aureleoules)

<details>
<summary>git diff</summary>

```diff
diff --git a/src/node/miner.cpp b/src/node/miner.cpp
index 0dc50c75c..c6838aee0 100644
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -395,7 +395,13 @@ void BlockAssembler::addPackageTxs(const CTxMemPool& mempool, int& nPackagesSele
             continue;
         }
 
-        auto ancestors{*Assert(mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false))};
+        auto ancestors_result{mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false)};
+        if (!ancestors_result) {
+            LogPrintLevel(BCLog::MEMPOOL, BCLog::Level::Error, 
+                          "%s: CalculateMemPoolAncestors failed unexpectedly, continuing with empty ancestor set (%s)",
+                          __func__, util::ErrorString(ancestors_result));
+        }
+        auto ancestors{ancestors_result.value_or(CTxMemPool::setEntries{})}; // for robustness, continue with empty ancestors if the calculation failed
 
         onlyUnconfirmed(ancestors);
         ancestors.insert(iter);
```
</details>

MarcoFalke replied:
At least you'll still need to use `Assume` to abort in debug builds

furszy replied:
> Interesting. I see your point re robustness, but generally, I really don't like silent failures - even when logged. I'd much rather they throw so we can more easily detect and fix unexpected behaviour as early on as possible

Well, at least on the mempool and validation areas, I'm on the totally opposite position.
If a result can have different meanings, then I would forcibly handle every possible outcome. We want robustness there.

Generally speaking: as the node's mempool state is built from the network, if there is a way to trigger something like this from an external source and we overseeing it (which could happen because we are humans), then an attacker could crash nodes remotely. And once that happens, there isn't a quick fix. Core isn't a common piece of software, any new version distribution and update coordination is hard.

Thus why I would rather be more conservative in terms of using assertions there. And instead, I would actually force us to not ignore errors paths (we could extend our result class to work similarly to the [Rust result type](https://doc.rust-lang.org/std/result/), so we are always forced to handle every possible outcome by default) or to rewrite code so it cannot fail in places that we don't expect failures.

This all comes from very bad experiences that I had in the past..

> At least you'll still need to use Assume to abort in debug builds

@MarcoFalke that wouldn't work, the result class internally asserts that the value exists when the `operator*()` is called (the upper `Assert()` call is actually redundant).

MarcoFalke replied:
> @MarcoFalke that wouldn't work, the result class internally asserts that the value exists when the operator*() is called (the upper Assert() call is actually redundant).

Sure, but not when `*` is replaced with `value_or`.

furszy replied:
> > @MarcoFalke that wouldn't work, the result class internally asserts that the value exists when the operator*() is called (the upper Assert() call is actually redundant).
> 
> Sure, but not when `*` is replaced with `value_or`.

Yeah, in that way we would retain the current behavior and add the assertion for debug builds only.

Still, future thing, I'm more inclined to rewrite the code so the function cannot fail or.., at least, add the error handler so it becomes clear what we do if it fails.

stickies-v replied:
> (the upper Assert() call is actually redundant)

I don't think that's true, Assert does a [boolean evaluation](https://github.com/bitcoin/bitcoin/blob/master/src/util/check.h#L66), so Result's `operator*()` [is not called](https://github.com/bitcoin/bitcoin/blob/3db23fd821e0a08e246a60c198676be42de5bcd6/src/util/result.h#L70).

> ... then an attacker could crash nodes remotely...

Point taken. I've reorganized (force pushing soon) the PR so that the first 2 commits keep the current (robust) behaviour but just using `util::Result`. In 2 subsequent commits I introduce and use a helper fn `AssumeCalculateMemPoolAncestors` to call `CalculateMemPoolAncestors` and always get a setEntries back, but for non-debug builds log the error in case of failure, or halt the application for debug builds. I think that strikes a nice balance between robustness in production and increasing visibility (of which currently we have 0 in master, for this issue) in development - hope you agree?

glozow replied:
Using `value_or` and `Assume()` seems like a good balance between handling every case and avoiding asserts in p2p-reachable codepaths. It's not great that some callsites ignore the return value of CMPA and continue using the ancestor set on master - I was going to comment about it until I saw the last couple commits. Agree adding an `Assume()` to catch in debug mode + logging is a better approach than halting in production.

stickies-v replied:
I think everyone's concerns are now addressed so I'll mark this as resolved, thanks for the input everyone.

stickies-v committed:
mempool: log/halt when CalculateMemPoolAncestors fails unexpectedly

When CalculateMemPoolAncestors fails unexpectedly (e.g. it exceeds
ancestor/descendant limits even though we expect no limits to be applied),
add an error log entry for increased visibility. For debug builds,
the application will even halt completely since this is not supposed
to happen.

stickies-v committed:
mempool: add AssumeCalculateMemPoolAncestors helper function

There are quite a few places that assume CalculateMemPoolAncestors
will return a value without raising an error. This helper function
adds logging (and Assume for debug builds) that ensures robustness
but increases visibility in case of unexpected failures

luke-jr commented:
If there's an actual bug, it should be fixed before/independently of the refactor (and only that fix backported).

stickies-v commented:
Force pushed to address review comments.

> If there's an actual bug, it should be fixed before/independently of the refactor (and only that fix backported)

I added [c25c6a6](https://github.com/bitcoin/bitcoin/pull/26289/commits/c25c6a6dc7843833d60da285d1681bfeb5022723) 

---

Most of the changes address [furszy's comments](https://github.com/bitcoin/bitcoin/pull/26289#discussion_r1001810274). In summary: ancestors/descendant limits unexpectedly getting hit no longer result in an assertion error (unless on debug builds) but in an error log message instead to avoid the node from crashing in case of a (non-fatal) unexpected error.

stickies-v commented:
Marking as draft until I've resolved the CI issues.

stickies-v commented:
PR is now ready for re-review - CI issues fixed. Looks like the [argument forwarding](https://github.com/bitcoin/bitcoin/compare/c7f82366f5bbde62f8cbd2511cf5af99b2e3860d..2f832f0c45aae3b51353db72691b4bc6450a7e9a#diff-c065d4cd2398ad0dbcef393c5dfc53f465bf44723348892395fffd2fb3bac522L296-R298) was causing issues on some platforms (ARM, Win64) - so I've just updated it to not use forwarding, since debugging it was going too slow and this approach is clean enough.

aureleoules reviewed (approved):
reACK 2f832f0c45aae3b51353db72691b4bc6450a7e9a

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -266,6 +266,7 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry,
                                            std::string &errString,
                                            bool fSearchForParents /* = true */) const
 {
+    setAncestors.clear();
```
can you explain why this is a bugfix? There is currently no code path that doesn't clear the set before calling this function.

MarcoFalke replied:
Ok, I see it is only in the package acceptance code path?

stickies-v replied:
> it is only in the package acceptance code path?

That's my understanding, yes. `MemPoolAccept::PreChecks` and `MemPoolAccept::SubmitPackage` are the only functions where setAncestors isn't (always) cleared before calling `CalculateMemPoolAncestors`. In `MemPoolAccept::AcceptMultipleTransactions()`, we first call `PreChecks()` and then `SubmitPackage()` with the same `Workspace`. `PreChecks` always seems to be called with a fresh `Workspace`, so I'd say `SubmitPackage` is the only issue. Given that this is currently only used in `testmempoolaccept` and `submitpackage` RPCs, it shouldn't be a particularly impactful bug.

I'll update the PR description a bit more now, and if I have to retouch, I'll update the bugfix commit message to add some more background.

glozow replied:
Correct, the only path is in `SubmitPackage`. It's not problematic to pass in `ws.m_ancestors` that is already populated, but I agree CMPA *should* always start with an empty ancestor set. Feel free to call it a bug or unexpected behavior etc.

stickies-v replied:
I wrote a quick check on `master` to verify that ancestors passed into CMPA are indeed always a subset of what is returned at the end of CMPA. At least as far as the scenarios covered in the unit and functional tests, that seems to be the case, since with this patch all tests complete without throwing.

<details>
<summary>git diff on master (5974c49f9)</summary>

```diff
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index 84ed2e9ef..e024ebe81 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -190,6 +190,8 @@ bool CTxMemPool::CalculateAncestorsAndCheckLimits(size_t entry_size,
                                                   const Limits& limits,
                                                   std::string &errString) const
 {
+    setEntries incoming_ancestors{setAncestors.begin(), setAncestors.end()};
+    setAncestors.clear();
     size_t totalSizeWithAncestors = entry_size;
 
     while (!staged_ancestors.empty()) {
@@ -226,6 +228,9 @@ bool CTxMemPool::CalculateAncestorsAndCheckLimits(size_t entry_size,
         }
     }
 
+    for (auto& ancestor : incoming_ancestors){
+        if (setAncestors.find(ancestor) == setAncestors.end()) throw std::runtime_error("incoming ancestor not found");
+    }
     return true;
 }
 ```
</details>

willcl-ark replied:
Why is this bugfix added here and then removed later?

stickies-v replied:
I'm unsure if we should keep the bugfix commit. As glozow says, the interface was brittle and could have been used in a buggy way, but that currently doesn't appear to be the case (as per my own verification too). With that understanding, we probably don't want to backport this? Alternatively, the separate bugfix commit could make it easier for downstream projects to patch this (potential) issue without having to adopt the full changeset? But I'm not sure that's how we usually do things.

cc @luke-jr - any further thoughts on this after [your earlier comment](https://github.com/bitcoin/bitcoin/pull/26289#issuecomment-1290725923) and [glozow's explanation](https://github.com/bitcoin/bitcoin/pull/26289#discussion_r1007899547)?

stickies-v commented on src/txmempool.cpp:
```diff
@@ -17,8 +17,10 @@
 #include <util/check.h>
 #include <util/moneystr.h>
 #include <util/overflow.h>
+#include <util/result.h>
```
Not necessary, but [developer notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#source-code-organization) state:

> Every .cpp and .h file should #include every header file it directly uses classes, functions or other definitions from, even if those headers are already included indirectly through other headers.

I'm not sure if just listing it as a return type classifies as "uses", but either way it seems like a sane thing to do?

glozow replied:
Oops, good point :)

JeremyRubin commented:
Would their be any merit to an alternative approach of logging how many current ancestors were passed in, and then subtracting that for the invariant check? That way you allocate less memory in the package code?

glozow commented:
> Would there be any merit to an alternative approach of logging how many current ancestors were passed in, and then subtracting that for the invariant check? That way you allocate less memory in the package code?

Having trouble understanding the question, can you elaborate on where the subtraction would be and which memory allocation you're referring to?

stickies-v commented:
Rebased to address the merge conflict from - no other changes:

```
git range-diff 85892f77c98c7a08834a06d52af3eb474275afd8 2f832f0 1f89af219a89da155d5302ca4fdc4f8f1f8612aa
```

@JeremyRubin I'm not sure I understand your question either, would appreciate if you could elaborate a bit? Thanks!

JeremyRubin commented:
I think I was mostly observing that the ancestors tracking is mostly being used for limits checking here, so there is an opportunity to (I think) not actually build a ancestor set and just use epochs to count the number of distinct ancestors directly. You can decrease the limit by another ancestor count if necessary.

stickies-v commented:
Thanks for clearing it up, you're talking about `CTxMemPoolEntry::Parents staged_ancestors`, right? @glozow suggested me that improved (offline) earlier as well, I agree that it looks like that can be improved with an epoch approach as you say. I've started working on it but not yet ready for PR. I'd say it's pretty orthogonal to the changes proposed here (we don't really touch `staged_ancestors`, focus is on the returned set of ancestors), so in my view it shouldn't be a hold up for making progress here, if you agree?

JeremyRubin commented:
sort of orthogonal, sort of not. if you do the refactor mentioned, i think you would end up not needing this refactor, so the two overlap.

stickies-v commented:
> i think you would end up not needing this refactor, so the two overlap.

I don't think I agree. The epoch refactoring would mostly target/remove `staged_ancestors`, whereas this PR targets the `setAncestors` that is passed into `CalculateMemPoolAncestors` as an `in,out` parameter. I don't see how using epoch affects `CalculateMemPoolAncestors` properly returning the ancestors as opposed adding them to the `setAncestors` reference?

JeremyRubin commented:
i'd need to look more detailed into the specifics, but no, the epoch patches could eliminate setAncestors and not the staged ancestors. Staging is still required since it is the work queue.

They way to look at it is in terms of where the setAncestors is consumed -- afaict, it's eventually transformed into just a count, so it's possible we can just keep the count directly if we use epochs to serve as a de-facto set. We still require staging to track what to expand next.

stickies-v commented:
For example in [`CTxMemPool::PrioritiseTransaction`](https://github.com/bitcoin/bitcoin/blob/master/src/txmempool.cpp#L863-L865) and [`CTxMemPool::UpdateForRemoveFromMempool`](https://github.com/bitcoin/bitcoin/blob/master/src/txmempool.cpp#L345-L348) we actually use the calculated ancestors in a way that doesn't seem trivial to refactor out.

willcl-ark reviewed (approved):
ACK 1f89af219a89da155d5302ca4fdc4f8f1f8612aa

I am unsure about the benefit of the first commit which is later removed, but it seems this might still be up for discussion.

The larger refactor related to removing out params and using Result looks good.

glozow commented on src/validation.cpp:
```diff
@@ -892,14 +890,15 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-            return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too-long-mempool-chain", errString);
-        }
+
+        auto invalid_too_long{state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too-long-mempool-chain", util::ErrorString(ancestors).original)};
+        if (ws.m_vsize > EXTRA_DESCENDANT_TX_SIZE_LIMIT) return invalid_too_long;
+        ancestors = m_pool.CalculateMemPoolAncestors(*entry, cpfp_carve_out_limits);
+        if (!ancestors) return invalid_too_long;
```
To reproduce - we probably can't test this since we can't call `PreChecks()` in isolation, but if you add `Assume(ws.m_state.IsValid())` right before `addUnchecked()` and run test/functional/mempool_package_onemore.py it will it.

stickies-v reviewed (commented):
Force pushed to [remove the bugfix commit ](https://github.com/bitcoin/bitcoin/pull/26289#discussion_r1034861371) and [address review comments](https://github.com/bitcoin/bitcoin/pull/26289#discussion_r1045931222) re instantiating `invalid_too_long` making the state invalid even when the failure path isn't hit

aureleoules commented:
CI is red @stickies-v

stickies-v commented:
> CI is red @stickies-v

Sorry, fixed now by capturing the original error message in a variable, so we don't show the error message from trying the cpfp carve-out.

stickies-v commented on src/node/miner.cpp:
```diff
@@ -394,9 +394,8 @@ void BlockAssembler::addPackageTxs(const CTxMemPool& mempool, int& nPackagesSele
-        std::string dummy;
-        mempool.CalculateMemPoolAncestors(*iter, ancestors, CTxMemPool::Limits::NoLimits(), dummy, false);
+        auto ancestors_result{mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false)};
+        auto ancestors{ancestors_result.value_or(CTxMemPool::setEntries{})};
```
Great question - and it pushed me quite deep down an interesting rabbit hole. I hadn't considered this. I think you're right that (without compiler optimization*) this (and all the other similar instances in this PR) would lead to a copy of the set, because we're calling `value_or()` on an lvalue so the [`const&` qualified `value_or()` method](https://github.com/bitcoin/bitcoin/blob/968f03e65c98a9a36aea6c57a29306839b0ada57/src/util/result.h#L60-L64) is used, which returns by value and doesn't benefit from move semantics.

By turning `ancestors_result` into an rvalue ref with `std::move`, we can benefit from the move semantics that `std::set` has implemented since now we'll be using the [`&&` qualified `value_or()` method](https://github.com/bitcoin/bitcoin/blob/968f03e65c98a9a36aea6c57a29306839b0ada57/src/util/result.h#L65-L69).

I've force pushed a change to replace all instances of `ancestors_result` into `std::move(ancestors_result)` as well as update `AssumeCalculateMemPoolAncestors()` to be explicit about using move semantics. _(note: most of these changes are undone in a later commit where we start using the newly introduced `AssumeCalculateMemPoolAncestors()` function)_

*I think, in practice, in most of the affected instances in this PR the copy operation actually would have been optimized away by the compiler (because I think `ancestors_result` just would have been inlined anyway, making them rvalues), but at the very least the code is now more explicit about intent, and potentially more performant - and I could be wrong about my optimization assumptions.

Thank you for raising this!

hebasto replied:
I'm late to this party... If the `T&& value()` member function could be added to the `Result` class, the following code will use a move assignment:
```c++
        auto ancestors_result{mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false)};
        CTxMemPool::setEntries ancestors;
        if (ancestors_result.has_value()) {
            ancestors = ancestors_result.value();
        }
```

stickies-v commented:
Force pushed to avoid set copy operations when calling `value_or()` on results, see [this comment for more background](https://github.com/bitcoin/bitcoin/pull/26289#discussion_r1047365141)

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26289/commits/47c4b1f52ab8d95d7deef83050bad49d1e3e5990

glozow reviewed (commented):
ACK 47c4b1f52ab8d95d7deef83050bad49d1e3e5990
code review to verify no change in behavior, and bench via ComplexMemPool and #26695 suggests no change in performance

stickies-v commented on src/validation.cpp:
```diff
@@ -1114,15 +1115,18 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
-            package_state.Invalid(PackageValidationResult::PCKG_MEMPOOL_ERROR,
-                                  strprintf("BUG! Mempool ancestors or descendants were underestimated: %s",
-                                            ws.m_ptx->GetHash().ToString()));
+        {
```
Because we're using move semantics in https://github.com/bitcoin/bitcoin/pull/26289/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R1129, so I wanted to be explicit about having `ancestors` go out of scope right after. I think that's worth the extra indentation?

furszy commented on src/txmempool.cpp:
```diff
@@ -254,9 +250,22 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry,
+                                            limits);
+}
+
+CTxMemPool::setEntries CTxMemPool::AssumeCalculateMemPoolAncestors(
+    std::string_view calling_fn_name,
+    const CTxMemPoolEntry &entry,
+    const Limits& limits,
```
Non-blocking point:

Could remove the `limits` arg. All the callers use `CTxMemPool::Limits::NoLimits()` and the only way `CalculateMemPoolAncestors` could fail is due a limit restriction.

furszy commented on src/txmempool.cpp:
```diff
@@ -254,9 +250,22 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry,
+                                            limits);
+}
+
+CTxMemPool::setEntries CTxMemPool::AssumeCalculateMemPoolAncestors(
+    std::string_view calling_fn_name,
```
Non-blocking nit:

This should be a helper/util static function rather than a class method. It's weird for a class method to receive the caller's function name just to log it internally.

furszy reviewed (commented):
light code review ACK 47c4b1f5

left two comments, nothing blocking.

aureleoules reviewed (approved):
ACK 47c4b1f52ab8d95d7deef83050bad49d1e3e5990
Great refactor that fixes a confusing function signature.

achow101 commented:
ACK 47c4b1f52ab8d95d7deef83050bad49d1e3e5990
